packages/dataviews/src/types.ts

```diff
@@ -757,11 +766,33 @@ export interface DataFormProps< Item > {
 	fields: Field< Item >[];
 	form: Form;
 	onChange: ( value: Record< string, any > ) => void;
+	/**
+	 * Validity information for form fields.
+	 */
+	validity?: FormValidity;
 }
 
 export interface FieldLayoutProps< Item > {
 	data: Item;
 	field: FormField;
 	onChange: ( value: any ) => void;
 	hideLabelFromVision?: boolean;
+	validity?: FieldValidity;
 }
+
+export interface FieldValidity {
+	id: string;
+	required?: 'invalid';
```

Comment:
I think there shouldn't be a special case for "required". required should also be `{ type: 'valid' | 'invalid' | 'validating', message }`

and I'd probably make this more generic. In other words

`Record< string, { type: 'valid' | 'invalid' | 'validating', message } >`

That way the same field and the same "custom" validation can generate multiple errors if needed.

Reply:
Update the Field validity so that every rule (required, elements, custom) has the same shape (object with type/message). This is not something we do internally: only `custom` is a function a Field author can use, but `required`/`elements` are calculated by ourselves synchronously. However, the validity object can be calculated by the consumer however they want, and the validating/valid states are possible in that scenario.